sql: integrate the name cache with the lease system; remove rename races

andreimatei opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6595)

<!-- Reviewable:end -->

andreimatei committed:
sql: integrate the table name cache with the leases

So here the problem: before this change, the planner had a cache used
to resolve table names to descriptor ids. This cache was updated by
gossip, async wrt the schema change transaction. This meant that the
name of a deleted table could be resolved to the old id for a while
after the drop committed. #5483

This change moves the cache into the leasing subsystem - a name is only
resolved to a table we have lease on, guaranteeing that the name is
valid.

This change in isolation makes things slightly worse than they were for
RENAME: RENAME is not a schema change; leases are not retracted when
we're about to perform a rename. RENAME does however trigger a gossip
update, which used to update the name cache. Now, the new cache is only
updated as a result of new descriptor versions being published, which is
not the case. The next commit makes RENAME be a proper schema change
that increments the descriptor version.

fixes #5483

andreimatei committed:
sql: make RENAME TABLE a schema change

Splitting the RENAME into a sync part (updating the descriptor) and an
async part (deleting the mapping of the old name to the id) makes it so
that the table name cache in LeaseManager can be reliably used to
resolve names: the schema changer waits until the old name is not in
use any more before removing the mapping, and the name cache is
guaranteed to have been updated when the transaction doing the RENAME
returns.

andreimatei commented:
I'll think of copacetic tests to add; wanted to see if this looks good early.

vivekmenezes commented:
Will take a look at it in the morning

vivekmenezes commented:
I'd like you to split this PR into two PRs. One that deals with the change to the lease manager and the other which applies the schema change. Is that possible?

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Will take a look at it in the morning</details>


---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/lease.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPuM9Fo_XesWFPX03e:-KHPuM9Fo_XesWFPX03f:654290395) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L444)):*
> ```Go
> 	mu sync.Mutex
> 	// The name of the table, as of the most recent lease.
> 	tableName string
> ```

Its unlikely to have more than a few names on the same table. I'd recommend just holding a slice of names where [0] is the latest name

---

*[sql/lease.go, line 463 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPvlFV00UhHParpGQz:-KHPvlFV00UhHParpGR-:-83396044) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L463)):*
> ```Go
> func (t *tableState) acquire(
> 	txn *client.Txn, version sqlbase.DescriptorVersion, store LeaseStore,
> ) (lease *LeaseState, err error) {
> ```

why named returned vars. Our style is against this unless it clearly brings value.

---

*[sql/sqlbase/structured.proto, line 282 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPyx8DK8Y03xeqEetR:-KHPyx8EzXyVWXlhW4ro:142120376) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/sqlbase/structured.proto#L282)):*
> ```protobuf
>   // descriptor before the rename.
>   optional uint32 old_parent_id = 20 [(gogoproto.nullable) = false];
>   optional string old_name = 21 [(gogoproto.nullable) = false];
> ```

Looks like you can't have two outstanding renames on a table. I think that is okay, but just checking.

Also looks like you are holding on to the old parent id because you can rename a table to be a part of another database?

Also these fields are all related to renaming, so it will be better to hold them in a transient protobuf that is only used during the transition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHPzlp7bMT9jduXZsx6:1038441105)*
<!-- Sent from Reviewable.io -->

tamird commented:
-1 on splitting this PR since the first commit breaks rename.

> leases are not retracted when we're about to perform a rename

What does it mean to retract a lease?

Also, can you add some tests to the first commit?



---

Reviewed 6 of 6 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 360 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ8BvT-XFmlBHqBwCm:-KHQ8BvT-XFmlBHqBwCn:1491940339) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/client/db_test.go#L360)):*
> ```Go
> 		key{txnType, "CommitInBatch"}:          {},
> 		key{txnType, "CommitOrCleanup"}:        {},
> 		key{txnType, "GetDeadline"}:            {},
> ```

this is in the wrong commit.

---

*[client/txn.go, line 425 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ5qWQxBdCh0T6UORG:-KHQ5qWRdiZu45cDEIEY:927984481) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/client/txn.go#L425)):*
> ```Go
> }
> 
> // GetDeadline sets the transactions deadline.
> ```

i dont think it does.

---

*[sql/internal.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6J_evYcWZ74VYJmf:-KHQ6J_f9ZhD0Xi_9q8X:1286602524) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/internal.go#L74)):*
> ```Go
> }
> 
> // getTableID retrieves the table ID for the specified table.
> ```

please move this code back; this movement only makes this harder to review.

---

*[sql/lease.go, line 680 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6fNFNp3l7kLIhgjb:-KHQ6fNFNp3l7kLIhgjc:-61131188) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L680)):*
> ```Go
> // Resolves a (database ID, table name) to the table descriptor's ID. Returns
> // false as the first component if the name is not part of the cache.
> func (t *tableNameCache) get(dbID sqlbase.ID, tableName string) (bool, sqlbase.ID) {
> ```

these return values should be reversed.

---

*[sql/lease.go, line 683 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6ky6xXHIv4Agt0D4:-KHQ6ky6xXHIv4Agt0D5:-2030414684) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L683)):*
> ```Go
> 	t.mu.Lock()
> 	defer t.mu.Unlock()
> 	if id, ok := t.tables[t.makeCacheKey(dbID, tableName)]; ok {
> ```

why not `return id, ok`?

---

*[sql/lease.go, line 695 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6arqsjLruOhY0MpD:-KHQ6arqsjLruOhY0MpE:-505040402) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L695)):*
> ```Go
> // cache.
> func (t *tableNameCache) upsert(
> 	dbID sqlbase.ID, oldName string, newName string, tableID sqlbase.ID) {
> ```

please follow the style guide.

---

*[sql/lease.go, line 711 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7-BB1rS3tQLGRmXS:-KHQ7-BB1rS3tQLGRmXT:190921904) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L711)):*
> ```Go
> 
> func (t *tableNameCache) makeCacheKey(dbID sqlbase.ID, tableName string) string {
> 	// TODO(andrei): Figure out an escaping scheme instead of this prefix.
> ```

use a struct as a cache key instead; you'll avoid allocating that way as well.

---

*[sql/lease.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7C-sEiXbfa0Jshk4:-KHQ7C-sEiXbfa0Jshk5:-862299652) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L726)):*
> ```Go
> 	// id; otherwise, the mapping may well be stale.
> 	// Not protected by mu.
> 	tableNames   tableNameCache
> ```

gotta be a pointer because mutex.

---

*[sql/lease.go, line 762 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7pMSuznthO4OLhpM:-KHQ7pMSuznthO4OLhpN:561806832) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L762)):*
> ```Go
> 	ok, tableID := m.tableNames.get(dbID, tableName)
> 	hadLease := ok && m.haveUsableLease(m.LeaseStore, tableID)
> 	if !ok || !hadLease {
> ```

`!ok` already implies `!hadLease`, you need not check both.

---

*[sql/lease.go, line 463 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPvlFV00UhHParpGQz:-KHQ6Ws3VtP3SdfCCrg1:5172523) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L463)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> why named returned vars. Our style is against this unless it clearly brings value.</details>
+1

---

*[sql/sqlbase/structured.go, line 671 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ8foUae132bvFK2O3:-KHQ8foUae132bvFK2O4:1117304910) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/sqlbase/structured.go#L671)):*
> ```Go
> }
> 
> // Visible returns true if the table is accessible to the general public.
> ```

as opposed to whom? this comment is misleading.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHQ6-tJp8FKhy0tKNJI:-1372301867)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Can you do it such it doesn't break rename. I'm trying to understand why adding the name into the leasing mechanism breaks rename?

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> -1 on splitting this PR since the first commit breaks rename.
> 
> > leases are not retracted when we're about to perform a rename
> 
> What does it mean to retract a lease?
> 
> Also, can you add some tests to the first commit?</details>


---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHQNpc0ewqNVaiNAIdB:1822088465)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: move getTableID to its only user - sql/internal

andreimatei commented:
They don't really make sense split, you need both of them at once. 
The lease manager changes are useless without the schema change (even worse, they break tests in isolation), because we're making the process of updating the name cache slower (it used to just take on gossip update, now it takes that update + acquiring some leases).
The schema change by itself doesn't accomplish any purpose (doesn't remove any races) since, well, nobody cares about the descriptor versions when resolving the names.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Can you do it such it doesn't break rename. I'm trying to understand why adding the name into the leasing mechanism breaks rename?</details>


---

Review status: 1 of 13 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[client/db_test.go, line 360 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ8BvT-XFmlBHqBwCm:-KHQCW69LPyBEz6YN00k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/client/db_test.go#L360)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is in the wrong commit.</details>
Done.

---

*[client/txn.go, line 425 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ5qWQxBdCh0T6UORG:-KHQDrgmosF4j-ylCKJr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/client/txn.go#L425)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i dont think it does.</details>
Done.

---

*[sql/internal.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6J_evYcWZ74VYJmf:-KHQBZOuKcHTIbPpdHyx:-716420180) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/internal.go#L74)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please move this code back; this movement only makes this harder to review.</details>
moved to a separate commit.

---

*[sql/lease.go, line 680 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6fNFNp3l7kLIhgjb:-KHQENZlKQSqhlaXeb4p:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L680)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> these return values should be reversed.</details>
Done.

---

*[sql/lease.go, line 683 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6ky6xXHIv4Agt0D4:-KHQENt30AsvlwXhujDg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L683)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not `return id, ok`?</details>
Done.

---

*[sql/lease.go, line 695 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ6arqsjLruOhY0MpD:-KHQE1w-ltEvclsI3XPP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L695)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please follow the style guide.</details>
Done.

---

*[sql/lease.go, line 711 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7-BB1rS3tQLGRmXS:-KHQFFwOPm_jhq6Eh1RZ:441343253) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L711)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> use a struct as a cache key instead; you'll avoid allocating that way as well.</details>
good idea

---

*[sql/lease.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7C-sEiXbfa0Jshk4:-KHQIB5ho2JcfTuqVZEw:-1273957387) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L726)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> gotta be a pointer because mutex.</details>
as discussed, this member is initialized once

---

*[sql/lease.go, line 762 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7pMSuznthO4OLhpM:-KHQFLJmgd7wLEQ3BqTQ:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L762)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `!ok` already implies `!hadLease`, you need not check both.</details>
done

---

*[sql/lease.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPuM9Fo_XesWFPX03e:-KHQ33N_3JtEalpd_r53:278154263) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L444)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Its unlikely to have more than a few names on the same table. I'd recommend just holding a slice of names where [0] is the latest name</details>
but I only need the last one (used to tell whether the name has been updated when we get a gossip update). 

---

*[sql/lease.go, line 463 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPvlFV00UhHParpGQz:-KHQ2z3vkVQDOM5XD8E3:2064267267) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L463)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> why named returned vars. Our style is against this unless it clearly brings value.</details>
yeah, leftover from using them in a defer. Removed.

---

*[sql/sqlbase/structured.go, line 671 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ8foUae132bvFK2O3:-KHQJ-xQQbFaPyv0VQ1u:-40656108) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/sqlbase/structured.go#L671)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> as opposed to whom? this comment is misleading.</details>
I got rid of this function

---

*[sql/sqlbase/structured.proto, line 282 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPyx8DK8Y03xeqEetR:-KHQKMVfT3tI4n18FIEv:-1840518743) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/sqlbase/structured.proto#L282)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Looks like you can't have two outstanding renames on a table. I think that is okay, but just checking.
> 
> Also looks like you are holding on to the old parent id because you can rename a table to be a part of another database?
> 
> Also these fields are all related to renaming, so it will be better to hold them in a transient protobuf that is only used during the transition.</details>
I got rid of the RENAME state. I made these old names a list so you can have multiple outstanding renames.

Added a comment to the parent_id.

Moved to a separate proto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHQ4DqTvWLkNXLxGvaC:-323689813)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: testing knobs improvements

- expose the LeaseManager through the testing shim
- expose the clock through the testing shim
- add a new notification type for rename schema changes

tamird commented:
Reviewed 10 of 10 files at r3, 8 of 8 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHREd6UWK3qtZWVLQU-:-KHREd6UWK3qtZWVLQU0:-769897411) ([raw file](https://github.com/cockroachdb/cockroach/blob/c459dc6600e86034f9f1aa8b977efeef311c0c1d/client/db_test.go#L359)):*
> ```Go
> 		key{txnType, "Commit"}:                 {},
> 		key{txnType, "CommitInBatch"}:          {},
> 		key{txnType, "GetDeadline"}:            {},
> ```

adding this twice, now?

---

*[sql/lease.go, line 711 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7-BB1rS3tQLGRmXS:-KHRDfi6lMIl-er45tPI:1368099729) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L711)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> good idea</details>
just get rid of this method, IMO

---

*[sql/lease.go, line 762 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7pMSuznthO4OLhpM:-KHRDppJk5mc16nz8vNO:112755512) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L762)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> done</details>
the local `hadLease` doesn't seem useful

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/sqlbase/structured.proto, line 277 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRQ_o_9tpUPoknqHFC:-KHRQ_o_9tpUPoknqHFD:1460709904) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L277)):*
> ```protobuf
>   optional State state = 19 [(gogoproto.nullable) = false];
>   // This array is populated if the descriptor was renamed. If the descriptor is
>   // renamed multiple times in a row, there can be more than one element.
> ```

renamed multiple times before the schema change is processed, there can be an element for each rename.

---

*[sql/sqlbase/structured.proto, line 283 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRPkupTqapsELtyxRr:-KHRPkupTqapsELtyxRs:1759254187) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L283)):*
> ```protobuf
> }
> 
> message TableRenameDetails {
> ```

Wanna embed this message inside the table descriptor making its type TableDescriptor_Rename

---

*[sql/sqlbase/structured.proto, line 286 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRPCr9jyHW9X5ikNHV:-KHRPCrAMKRt7ReRb7kE:-1000797804) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L286)):*
> ```protobuf
>   // The database that the table belonged to before the rename (tables can be
>   // renamed from one db to another).
>   optional uint32 old_parent_id = 1 [(gogoproto.nullable) = false];
> ```

s/old_parent_id/old_db/ and sets its type to ID and customname to OldDB

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/table.go, line 187 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRRrvHHDdhZ30FuBeE:-KHRRrvHHDdhZ30FuBeF:96139431) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/table.go#L187)):*
> ```Go
> 		curDeadline := p.txn.GetDeadline()
> 		if curDeadline == nil || time.Unix(0, curDeadline.WallTime).After(lease.Expiration()) {
> 			p.txn.SetDeadline(roachpb.Timestamp{WallTime: lease.Expiration().UnixNano()})
> ```

Another approach is to always call SetDeadline and have the txn handle that

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I'm still looking at this. These are the first comments

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> They don't really make sense split, you need both of them at once. 
> The lease manager changes are useless without the schema change (even worse, they break tests in isolation), because we're making the process of updating the name cache slower (it used to just take on gossip update, now it takes that update + acquiring some leases).
> The schema change by itself doesn't accomplish any purpose (doesn't remove any races) since, well, nobody cares about the descriptor versions when resolving the names.
> 
> </details>


---

Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 428 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRS4vjeBSfi_2eJ-Y2:-KHRS4vjeBSfi_2eJ-Y3:-398296489) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/client/txn.go#L428)):*
> ```Go
> func (txn *Txn) GetDeadline() *roachpb.Timestamp {
> 	return txn.deadline
> }
> ```

As mentioned later, let's not expose this API, instead expand SetDeadline to set the deadline or rename it to AdjustDeadline() and have AdjustDeadline pick the least value.

---

*[sql/internal.go, line 75 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRSXHotjCsuPjTaCDL:-KHRSXHotjCsuPjTaCDM:27871745) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/internal.go#L75)):*
> ```Go
> 
> // getTableID retrieves the table ID for the specified table.
> func getTableID(p *planner, qname *parser.QualifiedName) (sqlbase.ID, error) {
> ```

Since the planner already has a getDatabaseID method we might as well make this a method of the planner

---

*[sql/internal.go, line 87 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRT7JT1V45oYhWFA2r:-KHRT7JT1V45oYhWFA2s:368600374) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/internal.go#L87)):*
> ```Go
> 	nameKey := tableKey{dbID, qname.Table()}
> 	key := nameKey.Key()
> 	gr, err := p.txn.Get(key)
> ```

I didn't know you could redeclare err like this

---

*[sql/lease.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPuM9Fo_XesWFPX03e:-KHRTguE9ZzE79d7OLGM:1688700431) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L444)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> but I only need the last one (used to tell whether the name has been updated when we get a gossip update). </details>
I noticed that you had a tableNameCache above. I was wondering if you could replace this and the cache with a slice.

---

*[sql/lease.go, line 714 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRWmrHSvQZ6xP7nhc0:-KHRWmrHSvQZ6xP7nhc1:1102175092) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L714)):*
> ```Go
> 	}
> 	t.tables[t.makeCacheKey(dbID, newName)] = tableID
> }
> ```

delete(dbID, name, tableID)
insert(dbID, name, tableID)

Is there a strong need to create upsert?

---

*[sql/lease.go, line 732 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRXZ_vu0L5kDdyUZYi:-KHRXZ_vu0L5kDdyUZYj:214343908) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L732)):*
> ```Go
> 	// id; otherwise, the mapping may well be stale.
> 	// Not protected by mu.
> 	tableNames   tableNameCache
> ```

Wanna add a TODO to merge tableNames into tables above?

---

*[sql/lease.go, line 764 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRXzf1hAh7folU7apC:-KHRXzf1hAh7folU7apD:-994982800) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L764)):*
> ```Go
> // lease manager knows about.
> func (m *LeaseManager) AcquireByName(
> 	txn *client.Txn, dbID sqlbase.ID, tableName string,
> ```

I think we should create a new type

type tableFullName struct {
       dbID ID
       tableName string 
}

and use that everywhere in the code. Also, can we possibly use the db name instead of the db id?

---

*[sql/sqlbase/structured.proto, line 286 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRPCr9jyHW9X5ikNHV:-KHRZczu8BchtnrBzS9u:-1226331641) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L286)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> s/old_parent_id/old_db/ and sets its type to ID and customname to OldDB</details>
Do we really need to hold the db id instead of the db name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHRa5HUjC9gYIOxw7Pz:-1331867282)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: add more leasing and rename tests

andreimatei commented:
I've changed the design of the cache a bit: now its entries correspond directly to leases. IOW, while there's a lease with for name `foo`, there's gonna be a cache entry for that name. To make this clear, the cache now actually returns leases, and only still-valid leases. Before, the cache only kept name mappings, and it was updated through gossip, not through lease acquisition. I think the new version offers a better interface.
I've had to synchronize lease releasing by the `LeaseManager` (`TableState`) with returning leases from the cache. I've done this by adding synchronization state the `LeaseState`, in lack of a better idea.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'm still looking at this. These are the first comments</details>


---

Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHREd6UWK3qtZWVLQU-:-KHutTRIN_scQAIkS3Q_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c459dc6600e86034f9f1aa8b977efeef311c0c1d/client/db_test.go#L359)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> adding this twice, now?</details>
Done.

---

*[client/txn.go, line 428 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRS4vjeBSfi_2eJ-Y2:-KHusv7yNymoMK7Mohlr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/client/txn.go#L428)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> As mentioned later, let's not expose this API, instead expand SetDeadline to set the deadline or rename it to AdjustDeadline() and have AdjustDeadline pick the least value.</details>
Done.

---

*[sql/internal.go, line 75 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRSXHotjCsuPjTaCDL:-KHutq3GI_Sk4lhgQ4yA:1288291574) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/internal.go#L75)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Since the planner already has a getDatabaseID method we might as well make this a method of the planner</details>
well the difference is that, with tables, clients of the planner should get their ids from leases, which they get by name.

---

*[sql/internal.go, line 87 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRT7JT1V45oYhWFA2r:-KHus_J-HKkJOIXgasJH:683779746) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/internal.go#L87)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I didn't know you could redeclare err like this</details>
I believe the `:=` syntax, when outside of an `if`, just declares variables that are not already in scope. And it requires at least one such var.

---

*[sql/lease.go, line 711 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7-BB1rS3tQLGRmXS:-KHvRGvUvGnqVaucobtt:1486891174) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L711)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just get rid of this method, IMO</details>
I considered removing it, but it does the name normalization, so I think its reason enough to keep it.

---

*[sql/lease.go, line 762 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHQ7pMSuznthO4OLhpM:-KHvSBFt7FvcSDpG2pmc:677677860) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1386daccbb76ad7e97af05f07fea193cbb4798/sql/lease.go#L762)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the local `hadLease` doesn't seem useful</details>
gone

---

*[sql/lease.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHPuM9Fo_XesWFPX03e:-KHuu7I50iSSTvW2wOeZ:-462067546) ([raw file](https://github.com/cockroachdb/cockroach/blob/a740685b848cf388835748325a152b6a7c82ecf9/sql/lease.go#L444)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I noticed that you had a tableNameCache above. I was wondering if you could replace this and the cache with a slice.</details>
this field has gone away

---

*[sql/lease.go, line 714 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRWmrHSvQZ6xP7nhc0:-KHuuGx_6EJC9UEFYl-L:599620710) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L714)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> delete(dbID, name, tableID)
> insert(dbID, name, tableID)
> 
> Is there a strong need to create upsert?</details>
this interface has changed. There's no longer an upsert. I believe it made sense before for clients to not do deletions, but no longer.

---

*[sql/lease.go, line 732 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRXZ_vu0L5kDdyUZYi:-KHvTktL0nf-yzZxrCmM:-1055169533) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L732)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Wanna add a TODO to merge tableNames into tables above?</details>
not really :) How would the merge work? One is keyed by id and has granular locking, the other is keyed by name.
But in a sense the LeaseManager is the layer that does your merging - it hides descriptor ids from the clients.

---

*[sql/lease.go, line 764 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRXzf1hAh7folU7apC:-KHvRUyED9Gz9DrO0HpP:-1969844099) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/lease.go#L764)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I think we should create a new type
> 
> type tableFullName struct {
>        dbID ID
>        tableName string 
> }
> 
> and use that everywhere in the code. Also, can we possibly use the db name instead of the db id?</details>
I don't think it'd be useful for cache client for this type to be part of the interface. Instances would probably have to be created with a ctor that does the name normalization. I think it'd unnecessary burden.

About using the db name - the LeaseManager does deal with database names right now. Until we have a proper leasing story there, I think that's better kept out of the LeaseManager.

---

*[sql/table.go, line 187 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRRrvHHDdhZ30FuBeE:-KHusN2i_rArVX4sL3H6:-1755527534) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/table.go#L187)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Another approach is to always call SetDeadline and have the txn handle that</details>
done, and renamed to `UpdateDeadlineMaybe`

---

*[sql/sqlbase/structured.proto, line 277 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRQ_o_9tpUPoknqHFC:-KHvSf8pEZ2CDAQGz0rZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L277)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> renamed multiple times before the schema change is processed, there can be an element for each rename.</details>
Done.

---

*[sql/sqlbase/structured.proto, line 283 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRPkupTqapsELtyxRr:-KHvStZkgZ3vJt2_3Rf9:836221149) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L283)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Wanna embed this message inside the table descriptor making its type TableDescriptor_Rename</details>
I most certainly do

---

*[sql/sqlbase/structured.proto, line 286 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHRPCr9jyHW9X5ikNHV:-KHvT8ivFA2bHg9ZSbwl:1648963623) ([raw file](https://github.com/cockroachdb/cockroach/blob/03e6d3bff3340c76b8b0ae0e26fe6afa00f24d4f/sql/sqlbase/structured.proto#L286)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Do we really need to hold the db id instead of the db name?</details>
`parent_id` is the name of the descriptor field. We should rename both, if we do. But meh, what if we start having schemas?

And yeah, I think we need the id. Probably all the reasons why we have an db id and not a name in `parent_id` apply. Otherwise I don't know what would happen if these schema changes race with a db rename.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHvUU8LlHZkaTfy4R-J:712953033)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: fix a bug in purging old leases

Iterating over an aliased slice caused some leases to not be removed.

A test for this will be added in the later commit with the other tests
because it needs some methods in the LeaseManager that don't exist yet.

andreimatei commented:
@petermattis PTAL at r7, as you might have opinions about the `LeaseManager`

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I've changed the design of the cache a bit: now its entries correspond directly to leases. IOW, while there's a lease with for name `foo`, there's gonna be a cache entry for that name. To make this clear, the cache now actually returns leases, and only still-valid leases. Before, the cache only kept name mappings, and it was updated through gossip, not through lease acquisition. I think the new version offers a better interface.
> I've had to synchronize lease releasing by the `LeaseManager` (`TableState`) with returning leases from the cache. I've done this by adding synchronization state the `LeaseState`, in lack of a better idea.</details>


---

Review status: 0 of 21 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHw9gmtiDs-Gfo8Jdq2:2114272280)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r6.
Review status: 0 of 21 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[sql/lease.go, line 624 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHwBvldLaHWyBpIsSu1:-KHwBvleDNVQSx0sjQeW:1340306172) ([raw file](https://github.com/cockroachdb/cockroach/blob/3190545b5653d9eecd45a824861a5b94fc1d76e0/sql/lease.go#L624)):*
> ```Go
> 			} else {
> 				// Otherwise, all but the lease we just took are stale.
> 				toRelease = append([]*LeaseState(nil), t.active.data[:len(t.active.data)-1]...)
> ```

this change being in a commit by itself guaranatees that I can't review it. is this copy necessary because you're about to modify the slice? is the slice being modified out from under you? is it being appended to? are you appending to it?

There's a reason we prefer not to write commits with tests that follow later. If you need to expose some stuff to make the test possible, those changes should precede this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 21 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[sql/lease.go, line 624 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHwBvldLaHWyBpIsSu1:-KHzFgYWYu54uWgN3zEB:-107141816) ([raw file](https://github.com/cockroachdb/cockroach/blob/3190545b5653d9eecd45a824861a5b94fc1d76e0/sql/lease.go#L624)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this change being in a commit by itself guaranatees that I can't review it. is this copy necessary because you're about to modify the slice? is the slice being modified out from under you? is it being appended to? are you appending to it?
> 
> There's a reason we prefer not to write commits with tests that follow later. If you need to expose some stuff to make the test possible, those changes should precede this one.</details>
I messed up the split; there's a comment in `releaseLeasesIfNotActive` that explains what the bug was. Should be good now.
It's hard to get the reorder things for the test to come in here. The test would also not really be useful in understanding this commit - the test doesn't does fail because of this bug, but it's not suggestive of this particular problem. This class is not easy to unit-test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'm cool with you merging this in as is, but I think we should eventually break the ID based API such that you don't need to have methods like m.acquireFreshestFromStore(). If you feel the same you can create an issue for it.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> @petermattis PTAL at r7, as you might have opinions about the `LeaseManager`</details>


---

Review status: 0 of 21 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 371 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzOm9qXpaWe5O6QLRa:-KHzOm9rzWX7023WQXXY:-1138688158) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/client/db_test.go#L371)):*
> ```Go
> 		key{txnType, "SetSystemConfigTrigger"}: {},
> 		key{txnType, "SystemConfigTrigger"}:    {},
> 		key{txnType, "UpdateDeadlineMaybe"}:    {},
> ```

I'd call it either UpdateDeadline() or MaybeUpdateDeadline()

---

*[sql/lease.go, line 82 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzT19JTxBt4lRanPUC:-KHzT19K8Ix6gPN2Cs9y:-1702681684) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L82)):*
> ```Go
> // hasSomeLifeLeft returns true if the lease has at least a minimum of lifetime
> // left until expiration, and thus can be used.
> func (s *LeaseState) hasSomeLifeLeft(clock *hlc.Clock) bool {
> ```

s/hasSomeLifeLeft/retired/g

and use it as  !retired()

---

*[sql/lease.go, line 121 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzTuv0wxyMTUiTYw8l:-KHzTuv0wxyMTUiTYw8m:2116331142) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L121)):*
> ```Go
> 	lease := &LeaseState{}
> 	expiration := time.Unix(0, s.clock.Now().WallTime).Add(jitteredLeaseDuration())
> 	if !minExpirationTime.IsZero() && expiration.Before(minExpirationTime.Time) {
> ```

I don't think you need to check !IsZero() because a expiration will always be after zero.

---

*[sql/lease.go, line 514 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHz_MZz-0XKPsquXGlZ:-KHz_MZz-0XKPsquXGl_:-1063471858) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L514)):*
> ```Go
> 	lease *LeaseState, version sqlbase.DescriptorVersion, clock *hlc.Clock,
> ) *LeaseState {
> 	lease.mu.Lock()
> ```

Can this lock be moved further down just before recount++ ?

---

*[sql/lease.go, line 894 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzZ2ClwJQs0YQnzGD-:-KHzZ2ClwJQs0YQnzGD0:742085170) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L894)):*
> ```Go
> }
> 
> func nameMatchesLease(lease *LeaseState, dbID sqlbase.ID, tableName string) bool {
> ```

Pass a tableNameCacheKey?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KHzdZG3keemSdFgTuYf:-1656067560)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Not sure what you mean, exactly. One thing we should do about the ID-based API is make it private. The only reason why `Acquire` is still public is because of tests, but those can now be moved to the `sql` package.
`acquireFreshestFromStore` I don't really like, but I don't know how to get rid of. It's needed so that we can force getting a lease as recent as the name mapping that we've just read.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> :lgtm:
> 
> I'm cool with you merging this in as is, but I think we should eventually break the ID based API such that you don't need to have methods like m.acquireFreshestFromStore(). If you feel the same you can create an issue for it.</details>


---

Review status: 0 of 21 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[client/db_test.go, line 371 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzOm9qXpaWe5O6QLRa:-KI-3DqFvOzJXbl8_pDO:-1778224004) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/client/db_test.go#L371)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'd call it either UpdateDeadline() or MaybeUpdateDeadline()</details>
I like the `Maybe` as a suffix because it keeps the prefix of the name relevant...

---

*[sql/lease.go, line 82 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzT19JTxBt4lRanPUC:-KI-3Q_QjCv--Q5oN2ql:2021176684) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L82)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> s/hasSomeLifeLeft/retired/g
> 
> and use it as  !retired()</details>
but wouldn't `retired()` be misleading? The lease might still be active, just not with enough... life left in it.

---

*[sql/lease.go, line 121 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzTuv0wxyMTUiTYw8l:-KI-3v6xO6574_oseqke:265462694) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L121)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I don't think you need to check !IsZero() because a expiration will always be after zero.</details>
yeah... I've added it's apparent that zero value is accepted...

---

*[sql/lease.go, line 514 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHz_MZz-0XKPsquXGlZ:-KI-66PSWK0r20Bki26x:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L514)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Can this lock be moved further down just before recount++ ?</details>
done

---

*[sql/lease.go, line 894 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzZ2ClwJQs0YQnzGD-:-KI-4OFKN7K2cnmLCThm:2140986822) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L894)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Pass a tableNameCacheKey?</details>
well... tableNameCacheKey is now internal to the nameCache. At a minimum it'd need to be renamed so it can be used here, but I still don't think that's necessary. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KI--lmYXUnSaRacK3_S:-94024783)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I understand the need for it. Perhaps what I'm imagining is the refcount stuff being managed in the cache. Anyway Let's not do it here.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Not sure what you mean, exactly. One thing we should do about the ID-based API is make it private. The only reason why `Acquire` is still public is because of tests, but those can now be moved to the `sql` package.
> `acquireFreshestFromStore` I don't really like, but I don't know how to get rid of. It's needed so that we can force getting a lease as recent as the name mapping that we've just read.</details>


---

Review status: 0 of 21 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[sql/lease.go, line 82 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-KHzT19JTxBt4lRanPUC:-KI-KX3Vfq-_DWd5pDW8:-2039811341) ([raw file](https://github.com/cockroachdb/cockroach/blob/52e71ee72018d7e9dac2404b64ce2c078fda453a/sql/lease.go#L82)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> but wouldn't `retired()` be misleading? The lease might still be active, just not with enough... life left in it.</details>
notTakingNewRequests? I'm okay with what you have

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KI-M7-CWW_DvtHoLuvh:-1445066116)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@andreimatei Tried to look at r7, but reviewable isn't making that easy. Perhaps better to sit down and walk through the `LeaseManager` changes in person when I'm back next week. But don't let that hold up this PR.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I understand the need for it. Perhaps what I'm imagining is the refcount stuff being managed in the cache. Anyway Let's not do it here.</details>


---

Review status: 0 of 21 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6595#-:-KI-yamN5j9QKdW2tP8b:-1347736494)*
<!-- Sent from Reviewable.io -->
